bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp

```diff
@@ -231,3 +502,308 @@ vmcs_ut::test_launch_success()
+
+    mocks.OnCall(in.get(), intrinsics_intel_x64::vmread).With(VMCS_VIRTUAL_PROCESSOR_IDENTIFIER, _).Return(false);
+
+    RUN_UNITTEST_WITH_MOCKS(mocks, [&]
+    {
+        vmcs_intel_x64 vmcs(in);
+
+        EXPECT_EXCEPTION(vmcs.launch(host_state, guest_state), std::runtime_error);
+    });
+}
+
+void
+vmcs_ut::test_launch_print_execution_controls_failure()
+{
+    MockRepository mocks;
+    auto mm = mocks.Mock<memory_manager>();
+    auto in = bfn::mock_shared<intrinsics_intel_x64>(mocks);
+    auto host_state = bfn::mock_shared<vmcs_intel_x64_state>(mocks);
+    auto guest_state = bfn::mock_shared<vmcs_intel_x64_state>(mocks);
+
+    setup_vmcs_intrinsics(mocks, mm, in.get());
+    setup_vmcs_x64_state_intrinsics(mocks, host_state.get());
+    setup_vmcs_x64_state_intrinsics(mocks, guest_state.get());
+
+    mocks.OnCall(in.get(), intrinsics_intel_x64::vmlaunch).Return(false);
+    Call &read_host_rip = mocks.OnCall(in.get(), intrinsics_intel_x64::vmread).With(VMCS_HOST_RIP, _).Return(true);
+    mocks.OnCall(in.get(), intrinsics_intel_x64::vmread).After(read_host_rip).With(VMCS_PIN_BASED_VM_EXECUTION_CONTROLS, _).Return(false);
+
+    RUN_UNITTEST_WITH_MOCKS(mocks, [&]
+    {
+        vmcs_intel_x64 vmcs(in);
+
+        EXPECT_EXCEPTION(vmcs.launch(host_state, guest_state), std::runtime_error);
+    });
+}
+
+void vmcs_ut::test_launch_check_vmcs_control_state_failure()
+{
+    MockRepository mocks;
+    auto mm = mocks.Mock<memory_manager>();
+    auto in = bfn::mock_shared<intrinsics_intel_x64>(mocks);
+    auto host_state = bfn::mock_shared<vmcs_intel_x64_state>(mocks);
+    auto guest_state = bfn::mock_shared<vmcs_intel_x64_state>(mocks);
+
+    setup_vmcs_intrinsics(mocks, mm, in.get());
+    setup_vmcs_x64_state_intrinsics(mocks, host_state.get());
+    setup_vmcs_x64_state_intrinsics(mocks, guest_state.get());
+    setup_vmcs_launch_failure(mocks, in.get());
+
+    g_vmcs_fields[VMCS_CR3_TARGET_COUNT] = 5;
+
+    RUN_UNITTEST_WITH_MOCKS(mocks, [&]
+    {
+        vmcs_intel_x64 vmcs(in);
+
+        EXPECT_EXCEPTION(vmcs.launch(host_state, guest_state), std::logic_error);
+    });
+}
+
+void
+vmcs_ut::test_launch_check_guest_state_failure()
+{
+    MockRepository mocks;
+    auto mm = mocks.Mock<memory_manager>();
+    auto in = bfn::mock_shared<intrinsics_intel_x64>(mocks);
+    auto host_state = bfn::mock_shared<vmcs_intel_x64_state>(mocks);
+    auto guest_state = bfn::mock_shared<vmcs_intel_x64_state>(mocks);
+
+    setup_vmcs_intrinsics(mocks, mm, in.get());
+    setup_vmcs_x64_state_intrinsics(mocks, host_state.get());
+    setup_vmcs_x64_state_intrinsics(mocks, guest_state.get());
+    setup_vmcs_launch_failure(mocks, in.get());
+
+    g_vmcs_fields[VMCS_GUEST_CR0] = 0;
+
+    RUN_UNITTEST_WITH_MOCKS(mocks, [&]
+    {
+        vmcs_intel_x64 vmcs(in);
+
+        EXPECT_EXCEPTION(vmcs.launch(host_state, guest_state), std::logic_error);
+    });
+}
+
+void
+vmcs_ut::test_launch_check_host_state_failure()
```

Comment:
the check_xxx_state_failure functions can be removed. Your going to do this for every single check, so it's assumed that it's working. That being said... keep the code you have because your testing at least one of them in each of these.